hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SubscriptionAssignmentView.java

```diff
@@ -0,0 +1,204 @@
+        consumerNodeAssignments.putIfAbsent(nodeId, new HashSet<>());
+    }
+
+    private void addAssignment(SubscriptionAssignment assignment) {
+        subscriptionAssignments.get(assignment.getSubscriptionName()).add(assignment);
+        consumerNodeAssignments.get(assignment.getConsumerNodeId()).add(assignment);
+    }
+
+    private void removeAssignment(SubscriptionAssignment assignment) {
+        subscriptionAssignments.get(assignment.getSubscriptionName()).remove(assignment);
+        consumerNodeAssignments.get(assignment.getConsumerNodeId()).remove(assignment);
+    }
+
+    private void transferAssignment(String from, String to, SubscriptionName subscriptionName) {
+        removeAssignment(new SubscriptionAssignment(from, subscriptionName));
+        addAssignment(new SubscriptionAssignment(to, subscriptionName));
+    }
+
+    public SubscriptionAssignmentView deletions(SubscriptionAssignmentView target) {
+        return difference(this, target);
+    }
+
+    public SubscriptionAssignmentView additions(SubscriptionAssignmentView target) {
+        return difference(target, this);
+    }
+
+    private static SubscriptionAssignmentView difference(SubscriptionAssignmentView first, SubscriptionAssignmentView second) {
+        HashMap<SubscriptionName, Set<SubscriptionAssignment>> result = new HashMap<>();
+        for (SubscriptionName subscription : first.getSubscriptions()) {
+            Set<SubscriptionAssignment> assignments = first.getAssignmentsForSubscription(subscription);
+            if (!second.getSubscriptions().contains(subscription)) {
+                result.put(subscription, assignments);
+            } else {
+                Sets.SetView<SubscriptionAssignment> difference = Sets.difference(assignments, second.getAssignmentsForSubscription(subscription));
+                if (!difference.isEmpty()) {
+                    result.put(subscription, difference);
+                }
+            }
+        }
+        return new SubscriptionAssignmentView(result);
+    }
+
+    public static SubscriptionAssignmentView copyOf(SubscriptionAssignmentView currentState) {
+        return new SubscriptionAssignmentView(currentState.subscriptionAssignments);
+    }
+
+    public int getAssignmentsCountForSubscription(SubscriptionName subscription) {
+        return subscriptionAssignments.get(subscription).size();
+    }
+
+    public int getAssignmentsCountForConsumerNode(String nodeId) {
+        return consumerNodeAssignments.get(nodeId).size();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(subscriptionAssignments);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        SubscriptionAssignmentView that = (SubscriptionAssignmentView) o;
+        return Objects.equals(subscriptionAssignments, that.subscriptionAssignments);
```

Comment:
Yes, consumerNodeAssignments has the same contents as subscriptionAssignments but is differently indexed. SubscriptionAssignmentView is basically a MultiBiMap.